### PR TITLE
smooth_l1_loss

### LINKED ISSUE #21165: smooth_l1_loss
- [smooth_1_loss] #22378

### PR DESCRIPTION
**Pull request by sushmanthreddy**
close #21165

### CODE REVIEW

**Comment by AryanSharma21** [comment]
You can read about the tests in the documentations also 

https://unify.ai/docs/ivy/docs/ivy_tests.test_ivy.helpers.html

**Comment by sushmanthreddy** [comment]
@zaeemansari70 can u plss review this request and merge it?

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Hi @sushmanthreddy , I've fixed the small issue locally which I mentioned below to test your code. The failures are as follows.

```
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-numpy-False-False] - KeyError: 'cpu'
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-jax-False-False] - ivy.utils.exceptions.IvyBackendException: jax: smooth_l1_loss: jax: subtract: sub does not accept dtype bool at position 0. Accepted dtypes at position 0 are subtypes of complex...
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-torch-False-False] - ivy.utils.exceptions.IvyBackendException: torch: smooth_l1_loss: torch: subtract: Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` or `logical_...
FAILED ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py::test_smooth_l1_loss[cpu-paddle-False-False] - AssertionError
```
All the backends seems to be failing, kindly work on a fix for these after fixing the issue which I mentioned below, Thanks! :slightly_smiling_face:

**Comment by sushmanthreddy** [comment]
@zaeemansari70 can u pls review it once now??

**Comment by sushmanthreddy** [comment]
@AryanSharma21 @zaeemansari70 can u plss review this once?? here is the screen shoot of passing test cases

Screenshot from 2023-08-16 22-47-30


it is passing test cases in my local environment..can u pls review the pull request and give some feedback?

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Hey @sushmanthreddy ,
Your functional implementation seems to be failing for the falsifying example which I posted below:
```
E           ivy.utils.exceptions.IvyBackendException: paddle: smooth_l1_loss: paddle: where: (NotFound) The kernel with key (CPU, Undefined(AnyLayout), bfloat16) of kernel `where` is not registered. Selected wrong DataType `bfloat16`. Paddle support following DataTypes: int32, int64, float64, float32.
E             [Hint: Expected kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU != true, but received kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU:1 == true:1.] (at /paddle/paddle/phi/core/kernel_factory.cc:199)
E           
E           Falsifying example: test_smooth_l1_loss(
E               backend_fw='paddle',
E               on_device='cpu',
E               dtype_and_true=(['bfloat16'], [array([-1, -1, -1], dtype=bfloat16)]),
E               dtype_and_pred=(['bfloat16'], [array([-1, -1, -1], dtype=bfloat16)]),
E               beta=0.5,
E               reduction='none',
E               test_flags=FunctionTestFlags(
E                   ground_truth_backend='tensorflow',
E                   num_positional_args=2,
E                   with_out=False,
E                   instance_method=False,
E                   test_gradients=None,
E                   test_compile=None,
E                   as_variable=[False],
E                   native_arrays=[False],
E                   container=[False],
E               ),
E               fn_name='smooth_l1_loss',
E           )
E           
E           You can reproduce this example by temporarily adding @reproduce_failure('6.82.4', b'AXicY2BkAAMoBaZxsYlTdIABHQAAEPkAzg==') as a decorator on your test case
```

**Comment by sushmanthreddy** [comment]
@zaeemansari70 The stateful tests are failing because of a problem with the testing pipeline for the stateful API which as far as I know hasn't been fixed yet. The paddle test is failing because of an error with the bfloat16 datatype which again is a known issue that Ivy people are working on. 
Screenshot from 2023-08-17 00-36-58

see these are result of example loss functionswhich is failing for paddle,so pls review it once again

**Comment by sushmanthreddy** [comment]
@zaeemansari70 this issue is about adding backend and ivy  experimental functional api .So i have added the api  in ivy experimental folder and added the respectiive backend of it.And smooth_l1_loss is only supported in torch  and paddle .so I have added the backends and it is smooth_l1_loss is a mixed function .

And failure of tests, The stateful tests are failing because of a problem with the testing pipeline for the stateful API which as far as I know hasn't been fixed yet. The paddle test is failing because of an error with the bfloat16 datatype which again is a known issue that Ivy people are working on.
 
i hope this can be reviwed once again.I have provided the test cases screen shot above.so pls provide some feed backover it

File: ivy/functional/backends/paddle/experimental/losses.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    paddle.uint8,
+    paddle.float16,
+    paddle.complex64,
+    paddle.complex128,
+    paddle.bool,
+]
+
+
+@with_unsupported_device_and_dtypes(
+    {"2.5.1 and below": {"cpu": ("float16",)}}, backend_version
```
Comment (zaeemansari70) [inline ivy/functional/backends/paddle/experimental/losses.py:25]: The `unsupported_dtypes` will be added here in the decorator.
They should not be specified separately.
Kindly remove the list you added above.

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Requested some changes below. :+1: 

Also kindly uncommit all your `stateful` changes, the issue you linked to is a extension of the ivy API todolist.

Thanks!

File: ivy/functional/ivy/experimental/losses.py [hunk 2]
```diff
@@ -23,7 +24,7 @@ def log_poisson_loss(
     *,
     compute_full_loss: bool = False,
     axis: int = -1,
-    reduction: str = "none",
```
Comment (zaeemansari70) [inline ivy/functional/ivy/experimental/losses.py:26]: kindly uncommit this change too

File: ivy/functional/backends/torch/experimental/losses.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import torch
+import torch.nn.functional as F
```
Comment (zaeemansari70) [inline ivy/functional/backends/torch/experimental/losses.py:2]: You should just import `torch`

File: ivy/functional/backends/torch/experimental/losses.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    target: torch.Tensor,
+    /,
+    *,
+    beta: float = 1.0,
+    reduction: str = "mean",
+    size_average: Optional[bool] = None,
+    reduce: Optional[bool] = None,
+    out: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    return F.smooth_l1_loss(
```
Comment (zaeemansari70) [inline ivy/functional/backends/torch/experimental/losses.py:32]: and use it like this `torch.nn.functional.smooth_l1_loss()`

File: ivy/functional/backends/paddle/experimental/losses.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# global
+from typing import Optional
+import paddle
+import paddle.nn.functional as F
```
Comment (zaeemansari70) [inline ivy/functional/backends/paddle/experimental/losses.py:4]: same instructions for paddle as given below for torch

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
I've requested some changes 
Also please uncommit your `libssl` file which I assume you have mistakenly added.

Thanks!

File: ivy/functional/ivy/experimental/losses.py (hunk 2 above)
Comment (sushmanthreddy) [reply ivy/functional/ivy/experimental/losses.py:26]: this commit is about log_poission_loss..should i really change,these changes are made by someone..

**Comment by sushmanthreddy** [comment]
Screenshot from 2023-08-20 02-27-47 
checked supporting dtypes by this method.

File: ivy/functional/backends/paddle/experimental/losses.py (hunk 5 above)
Comment (zaeemansari70) [reply ivy/functional/backends/paddle/experimental/losses.py:4]: You haven't addressed this yet, you have to use paddle the same way you used torch

File: ivy/functional/ivy/experimental/losses.py (hunk 2 above)
Comment (zaeemansari70) [reply ivy/functional/ivy/experimental/losses.py:26]: The PR shows that you have updated the value of `reduction` to "mean", the only changes that should be shown should be just for `smooth_l1_loss` kindly un-commit the change which is being showed :slightly_smiling_face:

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
2 small changes which you missed to address before

**Comment by sushmanthreddy** [comment]
@zaeemansari70 i havee closed this  becuase of so many merge conflicts
i have made pull request ,if posssible plss assign to yourself and review it..they are basically same code from the last pr.
https://github.com/unifyai/ivy/pull/22378

**Pull request closed without merging**